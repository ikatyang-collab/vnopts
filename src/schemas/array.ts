import { Schema, SchemaHandlers } from '../schema';
import {
  DeprecatedResult,
  ExpectedResult,
  ForwardResult,
  NormalizedTransferResult,
  RedirectResult,
  Utils,
  ValidateResult,
} from '../types';

interface ArraySchemaParameters<$ValueSchema extends Schema<any>>
  extends SchemaHandlers<Array<$ValueSchema['_valueType']>> {
  name?: string;
  valueSchema: $ValueSchema;
}

export class ArraySchema<$ValueSchema extends Schema<any>> extends Schema<
  Array<$ValueSchema['_valueType']>,
  ArraySchemaParameters<$ValueSchema>
> {
  private _valueSchema: $ValueSchema;

  constructor({
    valueSchema,
    name = valueSchema.name,
    ...handlers
  }: ArraySchemaParameters<$ValueSchema>) {
    super({ ...handlers, name });
    this._valueSchema = valueSchema;
  }

  public expected(utils: Utils): ExpectedResult {
    const { text, list } = utils.normalizeExpectedResult(
      this._valueSchema.expected(utils),
    );

    return {
      text: text && `an array of ${text}`,
      list: list && {
        title: `an array of the following values`,
        values: [{ list }],
      },
    };
  }

  public validate(value: unknown, utils: Utils): ValidateResult {
    if (!Array.isArray(value)) {
      return false;
    }

    const invalidValues: Array<unknown> = [];

    for (const subValue of value) {
      const subValidateResult = utils.normalizeValidateResult(
        this._valueSchema.validate(subValue, utils),
        subValue,
      );
      if (subValidateResult !== true) {
        invalidValues.push(subValidateResult.value);
      }
    }

    return invalidValues.length === 0 ? true : { value: invalidValues };
  }

  public deprecated(
    value: this['_valueType'],
    utils: Utils,
  ): DeprecatedResult<this['_valueType']> {
    const deprecatedResult: DeprecatedResult<this['_valueType']> = [];

    for (const subValue of value) {
      const subDeprecatedResult = utils.normalizeDeprecatedResult(
        this._valueSchema.deprecated(subValue, utils),
        subValue,
      );

      if (subDeprecatedResult !== false) {
        deprecatedResult.push(
          ...subDeprecatedResult.map(({ value: deprecatedValue }) => ({
            value: [deprecatedValue],
          })),
        );
      }
    }

    return deprecatedResult;
  }

  public forward(
    value: this['_valueType'],
    utils: Utils,
  ): ForwardResult<this['_valueType']> {
    const forwardResult: ForwardResult<this['_valueType']> = [];

    for (const subValue of value) {
      const subForwardResult = utils.normalizeForwardResult(
        this._valueSchema.forward(subValue, utils),
        subValue,
      );

      forwardResult.push(...subForwardResult.map(wrapTransferResult));
    }

    return forwardResult;
  }

  public redirect(
    value: this['_valueType'],
    utils: Utils,
  ): RedirectResult<this['_valueType']> {
    const remain: this['_valueType'] = [];
    const redirect: ForwardResult<this['_valueType']> = [];

    for (const subValue of value) {
      const subRedirectResult = utils.normalizeRedirectResult(
        this._valueSchema.redirect(subValue, utils),
        subValue,
      );

      if ('remain' in subRedirectResult) {
        remain.push(subRedirectResult.remain!);
      }

      redirect.push(...subRedirectResult.redirect.map(wrapTransferResult));
    }

    return remain.length === 0 ? { redirect } : { redirect, remain };
  }

  public overlap(
    currentValue: this['_valueType'],
    newValue: this['_valueType'],
  ): this['_valueType'] {
    return currentValue.concat(newValue);
  }
}

function wrapTransferResult<$Value>(
  { from, to }: NormalizedTransferResult<$Value>, //
): NormalizedTransferResult<$Value[]> {
  return { from: [from], to };
}
