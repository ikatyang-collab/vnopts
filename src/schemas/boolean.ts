import { Schema, SchemaParameters } from '../schema.js'
import { ValidateResult } from '../types.js'

export class BooleanSchema extends Schema<boolean, SchemaParameters<boolean>> {
  public expected() {
    return 'true or false'
  }
  public validate(value: unknown): ValidateResult {
    return typeof value === 'boolean'
  }
}
