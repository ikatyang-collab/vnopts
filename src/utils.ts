import {
  DefaultResult,
  DeprecatedResult,
  ForwardResult,
  InvalidHandler,
  NormalizedDefaultResult,
  NormalizedDeprecatedResult,
  NormalizedDeprecatedResultWithTrue,
  NormalizedForwardResult,
  NormalizedInvalidHandler,
  NormalizedRedirectResult,
  NormalizedTransferResult,
  NormalizedValidateResult,
  Options,
  RedirectResult,
  TransferResult,
  ValidateResult,
} from './types';

export type NotEmptyArray<T> = [T, ...T[]];

export type RecordFromUnion<T extends Options, K extends keyof T> = {
  [X in T[K]]: Extract<T, Record<K, X>>
};

export function recordFromArray<
  T extends Record<string, any>,
  K extends keyof T
>(array: T[], mainKey: K): RecordFromUnion<T, K> {
  const record = Object.create(null) as Partial<RecordFromUnion<T, K>>;

  for (const value of array) {
    const key = value[mainKey];

    // istanbul ignore next
    if (record[key]) {
      throw new Error(`Duplicate ${mainKey} ${JSON.stringify(key)}`);
    }

    // @ts-ignore
    record[key] = value;
  }

  return record as RecordFromUnion<T, K>;
}

export function mapFromArray<T extends Record<string, any>, K extends keyof T>(
  array: T[],
  mainKey: K,
): Map<T[K], T> {
  const map = new Map<T[K], T>();

  for (const value of array) {
    const key = value[mainKey];

    // istanbul ignore next
    if (map.has(key)) {
      throw new Error(`Duplicate ${mainKey} ${JSON.stringify(key)}`);
    }

    map.set(key, value);
  }

  return map;
}

export function createAutoChecklist() {
  const map: { [idString: string]: boolean } = Object.create(null);
  return (id: any) => {
    const idString = JSON.stringify(id);
    if (map[idString]) {
      return true;
    }
    map[idString] = true;
    return false;
  };
}

export function partition<T>(
  array: T[],
  predicate: (x: T) => boolean,
): [T[], T[]] {
  const trueArray: T[] = [];
  const falseArray: T[] = [];

  for (const value of array) {
    if (predicate(value)) {
      trueArray.push(value);
    } else {
      falseArray.push(value);
    }
  }

  return [trueArray, falseArray];
}

export function isInt(value: number) {
  return value === Math.floor(value);
}

export function comparePrimitive(
  a: undefined | null | boolean | number | string,
  b: undefined | null | boolean | number | string,
): number {
  if (a === b) {
    return 0;
  }

  const typeofA = typeof a;
  const typeofB = typeof b;

  const orders = [
    'undefined',
    'object', // null
    'boolean',
    'number',
    'string',
  ];
  if (typeofA !== typeofB) {
    return orders.indexOf(typeofA) - orders.indexOf(typeofB);
  }

  if (typeofA !== 'string') {
    return Number(a) - Number(b);
  }

  return (a as string).localeCompare(b as string);
}

export function normalizeInvalidHandler(
  invalidHandler: InvalidHandler,
): NormalizedInvalidHandler {
  return (...args) => {
    const errorMessageOrError = invalidHandler(...args);
    return typeof errorMessageOrError === 'string'
      ? new Error(errorMessageOrError)
      : /* istanbul ignore next*/ errorMessageOrError;
  };
}

export function normalizeDefaultResult<$Value>(
  result: DefaultResult<$Value>,
): NormalizedDefaultResult<$Value> {
  return result === undefined ? {} : result;
}

export function normalizeValidateResult(
  result: ValidateResult,
  value: unknown,
): NormalizedValidateResult {
  return result === true ? true : result === false ? { value } : result;
}

export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
): NormalizedDeprecatedResult<$Value>;
export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
  doNotNormalizeTrue: true,
): NormalizedDeprecatedResultWithTrue<$Value>;
export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
  doNotNormalizeTrue = false,
):
  | NormalizedDeprecatedResult<$Value>
  | NormalizedDeprecatedResultWithTrue<$Value> {
  return result === false
    ? false
    : result === true
      ? doNotNormalizeTrue
        ? true
        : [{ value }]
      : 'value' in result
        ? [result]
        : result.length === 0
          ? false
          : (result as NotEmptyArray<{ value: $Value }>);
}

export function normalizeTransferResult<$Value>(
  result: TransferResult<$Value>,
  value: $Value,
): NormalizedTransferResult<$Value> {
  return typeof result === 'string' || 'key' in result
    ? { from: value, to: result }
    : 'from' in result
      ? { from: result.from!, to: result.to }
      : { from: value, to: result.to };
}

export function normalizeForwardResult<$Value>(
  result: ForwardResult<$Value>,
  value: $Value,
): NormalizedForwardResult<$Value> {
  return result === undefined
    ? []
    : Array.isArray(result)
      ? result.map(transferResult =>
          normalizeTransferResult(transferResult, value),
        )
      : [normalizeTransferResult(result, value)];
}

export function normalizeRedirectResult<$Value>(
  result: RedirectResult<$Value>,
  value: $Value,
): NormalizedRedirectResult<$Value> {
  const redirect = normalizeForwardResult(
    typeof result === 'object' && 'redirect' in result
      ? result.redirect
      : result,
    value,
  );
  return redirect.length === 0
    ? { remain: value, redirect }
    : typeof result === 'object' && 'remain' in result
      ? { remain: result.remain, redirect }
      : { redirect };
}
