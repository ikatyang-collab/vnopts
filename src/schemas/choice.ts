import { Schema, SchemaHandlers, SchemaParameters } from '../schema';
import {
  DeprecatedResult,
  ExpectedResult,
  ForwardResult,
  RedirectResult,
  TransferTo,
  Utils,
  ValidateResult,
} from '../types';
import { comparePrimitive, mapFromArray } from '../utils';

interface ChoiceInfo {
  value: ChoiceValue;
  deprecated?: boolean;
  redirect?: TransferTo;
  forward?: TransferTo;
  /** do not show this value in `expected` */
  hidden?: boolean;
}

type ChoiceValue = undefined | null | boolean | number | string;

interface ChoiceSchemaParameters
  extends SchemaParameters<ChoiceValue>,
    SchemaHandlers<ChoiceValue> {
  choices: Array<ChoiceValue | ChoiceInfo>;
}

export class ChoiceSchema extends Schema<ChoiceValue, ChoiceSchemaParameters> {
  private _choices: Map<ChoiceValue, ChoiceInfo>;

  constructor(parameters: ChoiceSchemaParameters) {
    super(parameters);
    this._choices = mapFromArray(
      parameters.choices.map(
        choice =>
          choice && typeof choice === 'object' ? choice : { value: choice },
      ),
      'value',
    );
  }

  public expected({ descriptor }: Utils): ExpectedResult {
    const choiceDescriptions = Array.from(this._choices.keys())
      .map(value => this._choices.get(value)!)
      .filter(({ hidden }) => !hidden)
      .map(choiceInfo => choiceInfo.value)
      .sort(comparePrimitive)
      .map(descriptor.value);

    const head = choiceDescriptions.slice(0, -2);
    const tail = choiceDescriptions.slice(-2);
    const message = head.concat(tail.join(' or ')).join(', ');

    return {
      text: message,
      list: {
        title: 'one of the following values',
        values: choiceDescriptions,
      },
    };
  }

  public validate(value: unknown): ValidateResult {
    return this._choices.has(value as any);
  }

  public deprecated(value: ChoiceValue): DeprecatedResult<ChoiceValue> {
    const choiceInfo = this._choices.get(value);
    return choiceInfo && choiceInfo.deprecated ? { value } : false;
  }

  public forward(value: ChoiceValue): ForwardResult<ChoiceValue> {
    const choiceInfo = this._choices.get(value);
    return choiceInfo ? choiceInfo.forward : undefined;
  }

  public redirect(value: ChoiceValue): RedirectResult<ChoiceValue> {
    const choiceInfo = this._choices.get(value);
    return choiceInfo ? choiceInfo.redirect : undefined;
  }
}
