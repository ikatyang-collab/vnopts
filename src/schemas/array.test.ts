import { eachHandler } from '../../tests/__helpers__/utils';
import { ArraySchema } from './array';
import { BooleanSchema } from './boolean';
import { ChoiceSchema } from './choice';

const name = '<key>';
const value = '<value>';
const invalid = '<invalid-value>';
const deprecated = '<deprecated-value>';
const transfer = '<transfer-value>';
const forward = '<forward-value>';
const redirect = '<redirect-value>';

describe.each`
  parameters | input                              | output                                    | hasWarnings
  ${{}}      | ${{ [name]: [value] }}             | ${{ [name]: [value] }}                    | ${false}
  ${{}}      | ${{ [name]: value }}               | ${Error}                                  | ${false}
  ${{}}      | ${{ [name]: [value, invalid] }}    | ${Error}                                  | ${false}
  ${{}}      | ${{ [name]: [value, deprecated] }} | ${{ [name]: [value, deprecated] }}        | ${true}
  ${{}}      | ${{ [name]: [value, forward] }}    | ${{ [name]: [value, forward, transfer] }} | ${false}
  ${{}}      | ${{ [name]: [value, redirect] }}   | ${{ [name]: [value, transfer] }}          | ${false}
  ${{}}      | ${{ [name]: [redirect] }}          | ${{ [name]: [transfer] }}                 | ${false}
`(
  '',
  eachHandler<ArraySchema<ChoiceSchema>>(ArraySchema, {
    valueSchema: new ChoiceSchema({
      name,
      choices: [
        value,
        { value: transfer },
        { value: redirect, redirect: { key: name, value: [transfer] } },
        { value: forward, forward: { key: name, value: [transfer] } },
        { value: deprecated, deprecated: true },
      ],
    }),
  }),
);

describe.each`
  parameters | input                    | output   | hasWarnings
  ${{}}      | ${{ [name]: 'invalid' }} | ${Error} | ${false}
`(
  '(simple expected values)',
  eachHandler<ArraySchema<BooleanSchema>>(ArraySchema, {
    valueSchema: new BooleanSchema({ name }),
  }),
);

describe.each`
  parameters | input                    | output   | hasWarnings
  ${{}}      | ${{ [name]: 'invalid' }} | ${Error} | ${false}
`(
  '(complex expected values)',
  eachHandler<ArraySchema<ChoiceSchema>>(ArraySchema, {
    valueSchema: new ChoiceSchema({
      name,
      choices: Array.from(new Array(20), (_value, index) => `value${index}`),
    }),
  }),
);
