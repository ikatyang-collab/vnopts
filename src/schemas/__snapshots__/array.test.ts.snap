// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(complex expected values) schema {"valueSchema":{"name":"<key>","_choices":{}}} with input {"<key>":"invalid"} 1`] = `
"Invalid \\"<key>\\" value. Expected an array of the following values, but received \\"invalid\\".
- one of the following values:
  - \\"value0\\"
  - \\"value1\\"
  - \\"value10\\"
  - \\"value11\\"
  - \\"value12\\"
  - \\"value13\\"
  - \\"value14\\"
  - \\"value15\\"
  - \\"value16\\"
  - \\"value17\\"
  - \\"value18\\"
  - \\"value19\\"
  - \\"value2\\"
  - \\"value3\\"
  - \\"value4\\"
  - \\"value5\\"
  - \\"value6\\"
  - \\"value7\\"
  - \\"value8\\"
  - \\"value9\\""
`;

exports[`(simple expected values) schema {"valueSchema":{"name":"<key>"}} with input {"<key>":"invalid"} 1`] = `"Invalid \\"<key>\\" value. Expected an array of true or false, but received \\"invalid\\"."`;

exports[`schema {"valueSchema":{"name":"<key>","_choices":{}}} with input {"<key>":"<value>"} 1`] = `
"Invalid \\"<key>\\" value. Expected an array of the following values, but received \\"<value>\\".
- one of the following values:
  - \\"<forward-value>\\"
  - \\"<redirect-value>\\"
  - \\"<transfer-value>\\"
  - \\"<value>\\""
`;

exports[`schema {"valueSchema":{"name":"<key>","_choices":{}}} with input {"<key>":["<value>","<deprecated-value>"]} 1`] = `
Array [
  "{ \\"<key>\\": [\\"<deprecated-value>\\"] } is deprecated.",
]
`;

exports[`schema {"valueSchema":{"name":"<key>","_choices":{}}} with input {"<key>":["<value>","<invalid-value>"]} 1`] = `
"Invalid \\"<key>\\" value. Expected an array of the following values, but received [\\"<invalid-value>\\"].
- one of the following values:
  - \\"<forward-value>\\"
  - \\"<redirect-value>\\"
  - \\"<transfer-value>\\"
  - \\"<value>\\""
`;
